# San Francisco employee Dataset analysis

Today In this project, I am going to analyze the San Francisco employee salary dataset using pandas DataFrame.The dataset is available in kaggle.


In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv(r"C:\Users\utsha\OneDrive\Desktop\Data analysis project\dataset\10. Salaries.csv")

C:\Users\utsha\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3457: DtypeWarning: Columns (3,4,5,6,12) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [3]:
df.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148049 non-null  object 
 4   OvertimePay       148654 non-null  object 
 5   OtherPay          148654 non-null  object 
 6   Benefits          112495 non-null  object 
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            38119 non-null   object 
dtypes: float64(3), int64(2), object(8)
memory usage: 14.7+ MB


In [5]:
df.shape

(148654, 13)

# Handling missing values

In [6]:
df.isnull().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

since columns notes and status has lots of missing value we are going to drop them.

In [7]:
df['Agency'].value_counts()

San Francisco    148654
Name: Agency, dtype: int64

agency also has one value which is San Francisco , so we are going to drop it too.

In [8]:
df.drop(columns = ['Notes' , 'Agency' , 'Status'] , axis = 1 ,inplace = True )
df.drop(columns = ['Id'] , axis = 1 ,inplace = True )

In [9]:
df.dtypes

EmployeeName         object
JobTitle             object
BasePay              object
OvertimePay          object
OtherPay             object
Benefits             object
TotalPay            float64
TotalPayBenefits    float64
Year                  int64
dtype: object

# Formatiing

In [10]:
df.head()

EmployeeName                                        JobTitle  \
0     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  
0  2011  
1  2011  
2  2011  
3  2011  
4  2011

In [11]:
df.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [12]:
df['BasePay'] = df.BasePay.replace('Not Provided','0')

In [13]:
df.BasePay = df.BasePay.astype('float')

In [14]:
avg_BasePay = df.BasePay.mean()



In [15]:
df['BasePay'] = df['BasePay'].replace({np.nan:avg_BasePay}) 

In [16]:

df['OvertimePay'] = df.OvertimePay.replace('Not Provided','0') 
df['OvertimePay'] = df['OvertimePay'].astype('float')


In [17]:
avg_OvertimePay = df.OvertimePay.mean()

In [18]:
df['OvertimePay'].fillna(avg_OvertimePay , inplace =True)

In [19]:
df.isnull().sum()

EmployeeName            0
JobTitle                0
BasePay                 0
OvertimePay             0
OtherPay                0
Benefits            36159
TotalPay                0
TotalPayBenefits        0
Year                    0
dtype: int64

In [20]:

df['Benefits'] = df.Benefits.replace('Not Provided' , '0')

In [21]:
df['Benefits'] = df['Benefits'].astype('float')

In [22]:
avg_Benefits = df['Benefits'].mean()
avg_Benefits

25007.003941775452

In [23]:
df['Benefits'].replace({np.nan:avg_Benefits},inplace = True)

In [24]:
df.isnull().sum()

EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
dtype: int64

In [25]:
df.dtypes

EmployeeName         object
JobTitle             object
BasePay             float64
OvertimePay         float64
OtherPay             object
Benefits            float64
TotalPay            float64
TotalPayBenefits    float64
Year                  int64
dtype: object

In [26]:
df.head()

EmployeeName                                        JobTitle  \
0     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay      Benefits   TotalPay  \
0  167411.18         0.00  400184.25  25007.003942  567595.43   
1  155966.02    245131.88  137811.38  25007.003942  538909.28   
2  212739.13    106088.18    16452.6  25007.003942  335279.91   
3   77916.00     56120.71   198306.9  25007.003942  332343.61   
4  134401.60      9737.00  182234.59  25007.003942  326373.19   

   TotalPayBenefits  Year  
0         567595.43  2011  
1         538909.28  2011  
2         335279.91  2011  
3         332343.61  2011  
4         326373.19  2011

In [27]:

df['OtherPay'] = df['OtherPay'].replace('Not Provided','0')
df['OtherPay'] = df['OtherPay'].astype('float')

In [28]:
df.OtherPay.dtypes

dtype('float64')

Now the data is completely cleaned and well formatted

## Time for some insights

In [29]:
df.describe()

BasePay    OvertimePay       OtherPay       Benefits  \
count  148654.000000  148654.000000  148654.000000  148654.000000   
mean    66323.656854    5065.923568    3648.669115   25007.003942   
std     42678.333515   11454.256594    8056.515703   13399.039858   
min      -166.010000      -0.010000   -7058.590000     -33.890000   
25%     33840.505000       0.000000       0.000000   22338.635000   
50%     65092.190000       0.000000     811.170000   25007.003942   
75%     94605.740000    4657.632500    4235.907500   33468.980000   
max    319275.010000  245131.880000  400184.250000   96570.660000   

            TotalPay  TotalPayBenefits           Year  
count  148654.000000     148654.000000  148654.000000  
mean    74768.321972      93692.554811    2012.522643  
std     50517.005274      62793.533483       1.117538  
min      -618.130000       -618.130000    2011.000000  
25%     36168.995000      44065.650000    2012.000000  
50%     71426.610000      92404.090000    2013.000000  
75%    105839.135000     132876.450000    2014.000000  
max    567595.430000     567595.430000    2014.000000

## Occurence of employee name (Top 5)

In [30]:
df.head(2)

EmployeeName                                        JobTitle    BasePay  \
0  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1    GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)  155966.02   

   OvertimePay   OtherPay      Benefits   TotalPay  TotalPayBenefits  Year  
0         0.00  400184.25  25007.003942  567595.43         567595.43  2011  
1    245131.88  137811.38  25007.003942  538909.28         538909.28  2011

In [31]:
df.EmployeeName.value_counts().head()

Kevin Lee       13
Richard Lee     11
Steven Lee      11
William Wong    11
Stanley Lee      9
Name: EmployeeName, dtype: int64

## Number of unique job titles

In [32]:
df['JobTitle'].nunique()

2159

There are 2159 unique jobs title

## Total Number of Job Titles Contain Captain

In [33]:
df.head(2)

EmployeeName                                        JobTitle    BasePay  \
0  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1    GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)  155966.02   

   OvertimePay   OtherPay      Benefits   TotalPay  TotalPayBenefits  Year  
0         0.00  400184.25  25007.003942  567595.43         567595.43  2011  
1    245131.88  137811.38  25007.003942  538909.28         538909.28  2011

In [34]:
len(df[df['JobTitle'].str.contains('Captain')])

411

411 job titles contain captain

## Display All the Employee Names From Fire Department

In [35]:
df[df['JobTitle'].str.contains('Fire Department')]['EmployeeName'].to_frame()

EmployeeName
36162     Joanne Hayes-White
72926   Joanne M Hayes-White
102303        Robert E Evans
110535  Joanne M Hayes-White

##  Find Minimum, Maximum, and Average BasePay

In [36]:
df.head(20)

EmployeeName                                          JobTitle  \
0       NATHANIEL FORD    GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1         GARY JIMENEZ                   CAPTAIN III (POLICE DEPARTMENT)   
2       ALBERT PARDINI                   CAPTAIN III (POLICE DEPARTMENT)   
3    CHRISTOPHER CHONG              WIRE ROPE CABLE MAINTENANCE MECHANIC   
4      PATRICK GARDNER      DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   
5       DAVID SULLIVAN                         ASSISTANT DEPUTY CHIEF II   
6            ALSON LEE                BATTALION CHIEF, (FIRE DEPARTMENT)   
7        DAVID KUSHNER                    DEPUTY DIRECTOR OF INVESTMENTS   
8       MICHAEL MORRIS                BATTALION CHIEF, (FIRE DEPARTMENT)   
9   JOANNE HAYES-WHITE            CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   
10       ARTHUR KENNEY  ASSISTANT CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   
11    PATRICIA JACKSON                   CAPTAIN III (POLICE DEPARTMENT)   
12   EDWARD HARRINGTON                       EXECUTIVE CONTRACT EMPLOYEE   
13         JOHN MARTIN                                 DEPARTMENT HEAD V   
14      DAVID FRANKLIN                BATTALION CHIEF, (FIRE DEPARTMENT)   
15     RICHARD CORRIEA                COMMANDER III, (POLICE DEPARTMENT)   
16            AMY HART                                 DEPARTMENT HEAD V   
17      SEBASTIAN WONG             CAPTAIN, EMERGENCYCY MEDICAL SERVICES   
18          MARTY ROSS                BATTALION CHIEF, (FIRE DEPARTMENT)   
19       ELLEN MOFFATT                        ASSISTANT MEDICAL EXAMINER   

      BasePay  OvertimePay   OtherPay      Benefits   TotalPay  \
0   167411.18         0.00  400184.25  25007.003942  567595.43   
1   155966.02    245131.88  137811.38  25007.003942  538909.28   
2   212739.13    106088.18   16452.60  25007.003942  335279.91   
3    77916.00     56120.71  198306.90  25007.003942  332343.61   
4   134401.60      9737.00  182234.59  25007.003942  326373.19   
5   118602.00      8601.00  189082.74  25007.003942  316285.74   
6    92492.01     89062.90  134426.14  25007.003942  315981.05   
7   256576.96         0.00   51322.50  25007.003942  307899.46   
8   176932.64     86362.68   40132.23  25007.003942  303427.55   
9   285262.00         0.00   17115.73  25007.003942  302377.73   
10  194999.39     71344.88   33149.90  25007.003942  299494.17   
11   99722.00     87082.62  110804.30  25007.003942  297608.92   
12  294580.02         0.00       0.00  25007.003942  294580.02   
13  271329.03         0.00   21342.59  25007.003942  292671.62   
14  174872.64     74050.30   37424.11  25007.003942  286347.05   
15  198778.01     73478.20   13957.65  25007.003942  286213.86   
16  268604.57         0.00   16115.86  25007.003942  284720.43   
17  140546.87    119397.26   18625.08  25007.003942  278569.21   
18  168692.63     69626.12   38115.47  25007.003942  276434.22   
19  257510.59       880.16   16159.50  25007.003942  274550.25   

    TotalPayBenefits  Year  
0          567595.43  2011  
1          538909.28  2011  
2          335279.91  2011  
3          332343.61  2011  
4          326373.19  2011  
5          316285.74  2011  
6          315981.05  2011  
7          307899.46  2011  
8          303427.55  2011  
9          302377.73  2011  
10         299494.17  2011  
11         297608.92  2011  
12         294580.02  2011  
13         292671.62  2011  
14         286347.05  2011  
15         286213.86  2011  
16         284720.43  2011  
17         278569.21  2011  
18         276434.22  2011  
19         274550.25  2011

In [37]:
# maximum basepay
df['BasePay'].max()

319275.01

In [38]:
# minimum basepay
df['BasePay'].min()

-166.01

In [39]:
# average basepay
df['BasePay'].mean()

66323.65685409939

# Replace 'Not Provided' in EmployeeName' Column to NaN 

In [40]:
df.head(2)

EmployeeName                                        JobTitle    BasePay  \
0  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1    GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)  155966.02   

   OvertimePay   OtherPay      Benefits   TotalPay  TotalPayBenefits  Year  
0         0.00  400184.25  25007.003942  567595.43         567595.43  2011  
1    245131.88  137811.38  25007.003942  538909.28         538909.28  2011

In [41]:
df.EmployeeName.replace({'Not provided':np.nan},inplace = True)

 Dropping rows with missing employee name The Rows Having 5 Missing Values

In [42]:
df.isnull().sum()

EmployeeName        4
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
dtype: int64

In [43]:
df.dropna(axis = 0 , inplace = True)

## Find Job Title of ALBERT PARDINI

In [44]:
df.head(2)

EmployeeName                                        JobTitle    BasePay  \
0  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1    GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)  155966.02   

   OvertimePay   OtherPay      Benefits   TotalPay  TotalPayBenefits  Year  
0         0.00  400184.25  25007.003942  567595.43         567595.43  2011  
1    245131.88  137811.38  25007.003942  538909.28         538909.28  2011

In [45]:
df[df['EmployeeName'].str.contains("ALBERT PARDINI")]['JobTitle']

2    CAPTAIN III (POLICE DEPARTMENT)
Name: JobTitle, dtype: object

## How Much ALBERT PARDINI Make (Include Benefits)?

In [46]:
df[df['EmployeeName'].str.contains("ALBERT PARDINI")]['TotalPayBenefits']

2    335279.91
Name: TotalPayBenefits, dtype: float64

## Display Name of The Person Having The Highest BasePay

In [47]:
df.sort_values(by = 'BasePay' , ascending = False)['EmployeeName']

72925           Gregory P Suhr
110532              Amy P Hart
72929            Robert L Shaw
72926     Joanne M Hayes-White
72931        Harlan L Kelly-Jr
                  ...         
72878          Richard Jackson
72874                Jordan Li
72872          Chung Huey Kung
72865             Robert Scott
72832           Irwin Sidharta
Name: EmployeeName, Length: 148650, dtype: object

# Find Average BasePay of All Employee Per Year 

In [48]:
df.groupby('Year')['BasePay'].mean()

Year
2011    63595.956517
2012    65436.406857
2013    69576.837750
2014    66564.421924
Name: BasePay, dtype: float64

## Find Average BasePay of All Employee Per JobTitle 

In [49]:
df.groupby('JobTitle')['BasePay'].mean().sort_values(ascending = False)

JobTitle
Chief of Police                            309767.683333
Chief, Fire Department                     304232.340000
Gen Mgr, Public Trnsp Dept                 297769.413333
CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)     285262.000000
Dep Dir for Investments, Ret               276153.765000
                                               ...      
PUBLIC SAFETY COMMUNICATIONS TECHNICIAN         0.000000
Conversion                                      0.000000
WAREHOUSE WORKER                                0.000000
RECREATION DIRECTOR                             0.000000
Cashier 3                                       0.000000
Name: BasePay, Length: 2158, dtype: float64

## Find Average BasePay of Employee Having Job Title ACCOUNTANT  

In [50]:
df[df['JobTitle'] == "ACCOUNTANT"]['BasePay'].mean()

46643.172

## Find Top 5 Most Common Jobs

In [51]:
df['JobTitle'].value_counts().head()

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64